src/crimson/os/poseidonstore/wal.h

```diff
@@ -0,0 +1,453 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "crimson/common/log.h"
+
+#include <boost/intrusive_ptr.hpp>
+
+#include <seastar/core/future.hh>
+
+#include "include/ceph_assert.h"
+#include "include/buffer.h"
+#include "include/denc.h"
+
+#include "crimson/osd/exceptions.h"
+#include "crimson/os/poseidonstore/poseidonstore_types.h"
+#include "crimson/os/poseidonstore/device_manager.h"
+#include <list>
+
+
+namespace crimson::os::poseidonstore {
+
+constexpr paddr_t WAL_START_ADDRESS = 0;
+constexpr uint32_t WAL_MAGIC = 0xCCCC;
+constexpr uint32_t WAL_DATA_HEADER_MAGIC = 0xBBBB;
+constexpr uint32_t WAL_COMMIT_HEADER_MAGIC = 0xFFFF;
+
+/**
+ * WAL in poseidonstore manages stream of small both data and metadata
+ */
+
+class WAL {
+public:
+  WAL(DeviceManager &device_manager);
+
+  using open_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  using open_ret = open_ertr::future<>; 
+  open_ret open();
+  //void close(); //TODO
+
+  using submit_entry_ertr = crimson::errorator<
+    crimson::ct_error::erange,
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::enospc
+    >;
+  using submit_entry_ret = submit_entry_ertr::future<
+    std::optional<std::pair<paddr_t, wal_seq_t>>
+    >;
+  /*
+   * submit_entry
+   *
+   * treat all bufferlists as a single write stream
+   * , and issue those bufferlists to the WAL.
+   *
+   * @param bufferlists to write
+   * @return <the start offset of entry submitted, 
+   * 	      wal_seq_t for the last entry submitted>
+   *
+   */
+  submit_entry_ret submit_entry(RecordRef rec, bool need_sync = true);
+
+  struct wal_data_block_header_t;
+  struct wal_header_t;
+  using write_block_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::erange>;
+  using write_block_ret = write_block_ertr::future<
+    std::pair<paddr_t, uint64_t>
+    >;
+  /*
+   * append_block
+   *
+   * append data to current write position of WAL
+   *
+   * Will write bl from offset to length
+   * 
+   * @param bufferlist to write
+   * @param laddr_t where data is written
+   * @return <relative address to write,
+   * 	      total length written to WAL (bytes)>
+   *
+   */
+  write_block_ret append_block(ceph::bufferlist bl, laddr_t addr);
```

Comment:
So, what does laddr_t mean for poseidon store?  poseidonstore_types.h mentions that it's the logical address from the pov of the onode, but that doesn't really make sense here.  Why would the wal know about that?  Which onode?